include/unifex/async_resource.hpp

```diff
@@ -0,0 +1,423 @@
+                        destroy_this_resource,   //
+                        deleter},
+        scheduler_(scheduler) {}
+
+  template <typename ResourceFactory>
+  auto construct(ResourceFactory&& factory) noexcept {
+    return unifex::on(
+        scheduler_,
+        unifex::just_from([factory = std::forward<ResourceFactory>(factory),
+                           this]() mutable noexcept {
+          // note for upstreaming: scope_ construction may succeed, but
+          // resource_ construction may fail
+          scope_.construct();
+          resource_.construct_with([&]() noexcept {
+            return factory(async_scope_ref{scope_.get(), this}, scheduler_);
+          });
+          return set_constructed();
+        }));
+  }
+
+  template <typename ResourceFactory>
+  auto construct_as_sender(ResourceFactory&& factory) noexcept {
+    return unifex::on(
+        scheduler_,
+        unifex::defer([factory = std::forward<ResourceFactory>(factory),
+                       this]() mutable noexcept {
+          // note for upstreaming: scope_ construction may succeed, but
+          // resource_ construction may fail
+          scope_.construct();
+          return unifex::then(
+              factory(async_scope_ref{scope_.get(), this}, scheduler_),
+              [&](auto&&... args) noexcept {
+                resource_.construct(static_cast<decltype(args)>(args)...);
+                return set_constructed();
+              });
+        }));
+  }
+
+  async_resource_ptr<Resource> ptr() noexcept {
+    return {&resource_.get(), &destructionEvt_};
+  }
+
+private:
+  _container* set_constructed() noexcept {
+    [[maybe_unused]] auto old =
+        state_.exchange(state_t::CONSTRUCTED, std::memory_order_seq_cst);
+    UNIFEX_ASSERT(old == state_t::DESTRUCT_SPAWNED);
+    return this;
+  }
+
+  UNIFEX_NO_UNIQUE_ADDRESS Scheduler scheduler_;
+  UNIFEX_NO_UNIQUE_ADDRESS unifex::manual_lifetime<Resource> resource_;
+
+  static void destruct_this_resource(_container_base* base) noexcept {
+    UNIFEX_ASSERT(base != nullptr);
+    auto& self = static_cast<_container&>(*base);
+    self.resource_.destruct();
+  }
+
+  // TODO potential improvmement to avoid the any_sender_of<> static method:
+  // 1. add operation states to the _container
+  // 2. maybe a union
+  static unifex::any_sender_of<>
+  schedule_destruct(_container_base* base, _container_base* parent) noexcept {
+    UNIFEX_ASSERT(base != nullptr);
+    auto& self = static_cast<_container&>(*base);
+    return unifex::sequence(
+        unifex::just_from([&self, parent]() noexcept {
+          [[maybe_unused]] auto old = self.state_.exchange(
+              state_t::DESTRUCT_SPAWNED, std::memory_order_seq_cst);
+          UNIFEX_ASSERT(old == state_t::ALLOCATED);
+          if (parent) {
+            parent->register_child(&self);
+          }
+        }),
+        unifex::unstoppable(unifex::on(self.scheduler_, self.destruct_impl())));
```

Comment:
This probably copies `scheduler_` so it might throw, plus `destruct_impl()`, maybe?